src/ray/core_worker/core_worker.cc

```diff
@@ -3016,40 +3016,52 @@ Status CoreWorker::ReportGeneratorItemReturns(
   RAY_LOG(DEBUG) << "Write the object ref stream, index: " << item_index
                  << ", id: " << dynamic_return_object.first;
 
-  waiter->IncrementObjectGenerated();
+  if (waiter) {
+    waiter->IncrementObjectGenerated();
+  }
+
   client->ReportGeneratorItemReturns(
       request,
       [waiter, generator_id, item_index](
           const Status &status, const rpc::ReportGeneratorItemReturnsReply &reply) {
         RAY_LOG(DEBUG) << "ReportGeneratorItemReturns replied. " << generator_id
                        << "index: " << item_index << ". total_consumed_reported: "
                        << reply.total_num_object_consumed();
-        RAY_CHECK(waiter != nullptr);
-        RAY_LOG(DEBUG) << "Total object consumed: " << waiter->TotalObjectConsumed()
-                       << ". Total object generated: " << waiter->TotalObjectGenerated();
-        if (status.ok()) {
-          /// Since unary gRPC requests are not ordered, it is possible the stale
-          /// total value can be replied. Since total object consumed only can
-          /// increment, we always choose the larger value here.
-          waiter->UpdateTotalObjectConsumed(
-              std::max(waiter->TotalObjectConsumed(), reply.total_num_object_consumed()));
-        } else {
-          // TODO(sang): Handle network error more gracefully.
-          // If the request fails, we should just resume until task finishes without
-          // backpressure.
-          waiter->UpdateTotalObjectConsumed(waiter->TotalObjectGenerated());
-          RAY_LOG(WARNING) << "Failed to send the object ref.";
+        if (waiter) {
```

Comment:
Should we avoid passing a callback entirely if there's no waiter? Not sure if this is handled any differently internally

Reply:
We still need to check for signals, waiter is just a back-pressure hook for us to hold down the thread if we want to slow down the producer